File: scripts/Serialize and Deserialize Binary Tree/Serialize and Deserialize Binary Tree.js
/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
 var serialize = function(root) {
  let result = [];
  const dfs = (r) => {
    if (!r) {
      result.push('N');
      return;
    }
    result.push(r.val);
    dfs(r.left);
    dfs(r.right);
  }
  dfs(root);
  return result.join(",")
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
  const vals = data.split(",");
  const dfs = (stream) => {
    const item = stream.shift();
    if (item == 'N') {
      return null;
    }
    const node = new TreeNode(item);
    node.left = dfs(stream);
    node.right = dfs(stream);
    return node;
  };

  return dfs(vals);
};

File: algorithms/S/Serialize and Deserialize Binary Tree/Serialize and Deserialize Binary Tree.java
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        String data="";
        Queue<TreeNode> q=new LinkedList<>();
        if(root!=null)
        q.add(root);
        else
         return "";
        data=Integer.toString(root.val)+"e";
        while(!q.isEmpty())
        {
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                TreeNode node=q.poll();
                if(node.left!=null)
                {
                    data=data+Integer.toString(node.left.val)+"e";
                    q.add(node.left);
                }
                else
                    data=data+"N"+"e";
                if(node.right!=null)
                {
                    data=data+Integer.toString(node.right.val)+"e";
                    q.add(node.right);
                }
                else
                    data=data+"N"+"e";
            }
        }
        return data;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.length()==0)
            return null;
        int i=0;
        String s="";
        while(data.charAt(i)!='e')
            s=s+data.charAt(i++);
        int d=Integer.parseInt(s);
        TreeNode root=new TreeNode(d);
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        while(i<data.length() && !q.isEmpty())
        {
            int size=q.size();
            for(int j=0;j<size;j++)
            {
                s="";
                i++;
                TreeNode node=q.poll();
                while(data.charAt(i)!='e')
                   s=s+data.charAt(i++);
                if(s.equals("N"))
                    node.left=null;
                else
                {
                    TreeNode l=new TreeNode(Integer.parseInt(s));
                    node.left=l;
                    q.add(l);
                }
                s="";
                i++;
                while(data.charAt(i)!='e')
                   s=s+data.charAt(i++);
                if(s.equals("N"))
                    node.right=null;
                else
                {
                    TreeNode r=new TreeNode(Integer.parseInt(s));
                    node.right=r;
                    q.add(r);
                }
            }
        }
        return root;
        
    }
}
